timemachine/md/smc.py

```diff
@@ -120,6 +122,158 @@ def accumulate_results(samples, indices, log_weights, incremental_log_weights):
+        "ancestry_traj"
+            [K-1, N] array of ancestor idxs
+        "log_weights_traj"
+            [K, N] array of accumulated log weights
+        "lambdas_traj"
+            [K] array of adaptive lambdas
+
+    References
+    ----------
+    * [Zhou, Johansen, Aston, 2016]
+        Towards Automatic Model Comparison: An Adaptive Sequential Monte Carlo Approach
+        https://arxiv.org/pdf/1303.3123 (Algorithm #4)
+    """
+    n = len(samples)
+
+    # check cess_target
+    assert cess_target > 1, f"cess_target is too small: {cess_target} <= 1"
+    assert cess_target < n, f"cess_target is too large: {cess_target} >= {n}"
+
+    log_weights = np.zeros(n)
+    norm_log_weights = log_weights - logsumexp(log_weights)
+
+    # store
+    sample_traj = [samples]
+    ancestry_traj = [np.arange(n)]
+    log_weights_traj = [np.array(log_weights)]
+    incremental_log_weights_traj = []  # note: redundant but convenient
+    lambdas_traj = [0.0]
+
+    def accumulate_results(samples, indices, log_weights, incremental_log_weights, lam_target):
+        if store_intermediate_traj:
+            sample_traj.append(samples)
+        else:
+            # only store one intermediate set of samples to reduce memory usage
+            sample_traj[0] = samples
+        ancestry_traj.append(indices)
+        log_weights_traj.append(np.array(log_weights))
+        incremental_log_weights_traj.append(np.array(incremental_log_weights))
+        lambdas_traj.append(lam_target)
+
+    lam_initial = 0.0
+    lam_target = 1.0  # adapted
+
+    # Main ASMC loop
+    while True:
+
+        # binary search for lambda that gives cess ~= cess_target
+        left_lambda = lam_initial
+        right_lambda = lam_target
+
+        cur_log_prob = log_prob(sample_traj[-1], lam_initial, True)
+        while True:
+            mid_lambda = max(min(lam_target, left_lambda + (right_lambda - left_lambda) / 2), lam_initial)
+            incremental_log_weights = log_prob(sample_traj[-1], mid_lambda, False) - cur_log_prob
+            cess = conditional_effective_sample_size(norm_log_weights, incremental_log_weights)
+            if left_lambda > right_lambda:
+                # Let the loop run above so we can sample at the lam_target exactly
+                break
+            if cess < cess_target:  # too small
+                right_lambda = mid_lambda - epsilon
+            elif cess > cess_target:  # too big
+                left_lambda = mid_lambda + epsilon
+            elif mid_lambda == lam_target:
+                break
+
+        if mid_lambda == lam_initial:
+            # must move at least epsilon
+            mid_lambda += epsilon
+
+        lam_target = mid_lambda
+
+        # Stop when lam_target == 1.0
+        #   See
+        #   * discussion at https://github.com/proteneer/timemachine/pull/718#discussion_r854276326
+        #   * helper function get_endstate_samples_from_smc_result
+        if lam_target == 1.0:
+            break
```

Comment:
could we simplify using scipy.optimize.root_scalar ? (as in https://github.com/proteneer/timemachine/pull/442/files#diff-cb96fac1a84532e2c8d063fdcb9f0695dff44d884211290d23817b0d7d57651cR94 )